Data Structures/Linked-Lists/list-implementation.cpp

```diff
@@ -0,0 +1,417 @@
+// Name: Harsh Aryan
+// Email-id: 12harsharyan@gmail.com
+// THIS IS A PROGRAM TO IMPLEMENT LIST . IT CONSISTS OF INSERTION, DELETION, TRAVERSAL, SEARCH, REVERSE.
+
+#include <stdio.h>
+#include <stdlib.h>
+
+//declaring  a structure named node
+//a node is the smallest unit of  a linked list
+
+struct node
+{
+    int data;
+    struct node *next;
+};
+
+//function to create a node
+//worst case time complexity O(1)
+//best case time complexity O(1)
+//these are just constant time operations
+// Space Complexity: O(1), Since we have just used an additional variable max inside the function.
+
+struct node *createNewNode(int d)
+{
+    struct node *temp;
+    temp=(struct node *)malloc(sizeof(struct node));
+    temp->next=NULL;
+    temp->data=d;
+    return temp;
+}
+
+//function to add a new node at the beginning of the list
+//worst case time complexity O(1)
+//best case time complexity O(1)
+//these are just constant time operations
+// Space Complexity: O(1), Since we have just used an additional variable max inside the function.
+
+
+struct node *addAtBeg(struct node *head, int d)
+{
+    struct node *temp;
+    temp=createNewNode(d);
+    if(head==NULL)
+    {
+        head=temp;
+        return head;
+    }
+    else
+    {
+        temp->next=head;
+        head=temp;
+        return head;
+    }
+}
+
+//function to add a new node at the end of the list
+//worst case time complexity O(N)
+//best case time complexity O(N)
+// Inside the for loop, there is just a condition check ,
+// Overall, time taken is O(n) thus.
+// Space Complexity: O(1), Since we have just used an additional variable max inside the function.
+
+struct node *addAtEnd(struct node *head, int d)
```

Comment:
You can add a pointer that keeps track of the end of the LL to make insertion at the end O(1) instead of O(n). This is recommended and is often asked in interviews. Keep this in mind.